frontend/src/views/instructor/ddahs/import.tsx

```diff
@@ -0,0 +1,319 @@
+import React from "react";
+import { Alert } from "react-bootstrap";
+import { FaArrowRight } from "react-icons/fa";
+import { useSelector } from "react-redux";
+import { assignmentsSelector, applicantsSelector } from "../../../api/actions";
+import { ddahsSelector, upsertDdahs } from "../../../api/actions/ddahs";
+import {
+    Ddah,
+    Assignment,
+    Applicant,
+    MinimalDdah,
+    RawDuty,
+} from "../../../api/defs/types";
+import { ImportActionButton } from "../../../components/import-button";
+import { DiffSpec, diffImport, getChanged } from "../../../libs/diffs";
+import { normalizeDdahImports } from "../../../libs/import-export";
+import { useThunkDispatch } from "../../../libs/thunk-dispatch";
+
+enum DiffType {
+    Unchanged = "UNCHANGED",
+    Created = "CREATED",
+    Updated = "UPDATED",
+    Deleted = "DELETED",
+}
+
+type MinimalDuty = {
+    description: string;
+    hours: number;
+};
+
+interface DdahUpdate {
+    updatedDdah: Ddah;
+    oldDdah: Ddah;
+}
+
+interface DutyDiff {
+    oldDuty?: MinimalDuty;
+    newDuty?: MinimalDuty;
+    status: DiffType;
+}
+
+function getDutiesDiff(oldDuties: RawDuty[], newDuties: RawDuty[]) {
+    let changes: DutyDiff[] = [];
+    let dutiesChanged = false;
+
+    oldDuties.forEach((oldDuty) => {
```

Comment:
Prefer for loops over `forEach`. They're much faster.

Reply:
So far they are because they are older, but `forEach` loops are easier to read in my opinion, especially when we don't need the iterator. And if we trade a tiny performance increase bound by the number of duties, I feel it would make sense if the code is easier to reason about. Let me know your thoughts!

Reply:
I disagree on both accounts. `forEach` loops are slower because they involve transferring execution to another function repeatedly. I also don't think they contribute to readability :-P